Provide Scopes value object to encapsulate scope operations

rezaansyed opened the pull request:
## What does this PR solve?
There are common operations that are used for scopes on OmniAuth and apps consuming the gem. Such operations include serializing, deserializing and normalizing scopes. This PR introduces the `Scopes` value object that will be used to perform such operations. This helps ensure that apps do not potentially write duplicate code for scopes and delegating this responsibility to the Shopify OmniAuth gem.

ragalie commented on lib/omniauth/strategies/shopify.rb:
```diff
@@ -74,16 +73,12 @@ def valid_signature?
-        scope_list - ignore_scopes
+        config = options.authorize_params.merge(options_for("authorize"))
+        return false unless token && config[:scope] && token['scope']
+        scope = config[:scope]
```
Maybe inline this?

ragalie commented on test/shopify/scopes_test.rb:
```diff
@@ -0,0 +1,15 @@
+require 'omniauth-shopify-oauth2'
+
+class ScopesTest < Minitest::Test
```
If folks are going to use this outside of the gem then I would fill out the unit tests here.

ragalie reviewed (commented):
I really like this!

Left some comments but onboard with the approach!

tanema commented on lib/omniauth/shopify/scopes.rb:
```diff
@@ -0,0 +1,45 @@
+module OmniAuth
+  module Shopify
+    class Scopes
+      SCOPE_DELIMITER = ','
```
This all seems overcomplicated when you are just wrapping a Set. We could just inherit from Set and normalize on deserialization.

```ruby
module OmniAuth
  module Shopify
    class Scopes < Set
      SCOPE_DELIMITER = ','

      def self.deserialize(scopes)
        scope_list = scopes
          .to_s
          .split(SCOPE_DELIMITER)
          .map(&:strip)
          .reject(&:empty?)
        ignore_scopes = scope_list
          .map { |scope| scope =~ /\A(unauthenticated_)?write_(.*)\z/ && "#{$1}read_#{$2}" }
          .compact
        new(scope_list - ignore_scopes)
      end

      def serialize
        to_a.join(SCOPE_DELIMITER)
      end
    end
  end
end
```

alexaitken replied:
I would be favour of this change if `normalize` is kept as a method on `Scopes` instead of building it into the deserialize method.  This will allow the caller to determine if they want a normalized value even if they don't have original serialized string.

rezaansyed replied:
I agree. I think it's best to leave `normalize` that is up to the caller on whether they need it. I just made the refactor on my end and that approach works.

mSitkovets commented on lib/omniauth/shopify/scopes.rb:
```diff
@@ -0,0 +1,48 @@
+      def normalize
+        ignore_scopes = scopes.map { |scope| scope =~ /\A(unauthenticated_)?write_(.*)\z/ && "#{$1}read_#{$2}" }.compact
+
+        Scopes.new(scopes - ignore_scopes)
```
Here you're explicitly writing out `Scopes.new` but above you only call `new` within the self.deserialize. Which way should we go with/which is the best practice? Do we only leave out the `Scopes` part if it's a class method?

rezaansyed replied:
We can call `new` in `self.deserialize` because `new` is a class method and it will call `initialize` under the hood. When creating a new object inside of an instance, then you have to access the class to make a call to the class method `new`.

tanema reviewed (commented):
Okay, so the code looks okay but I do not see how this makes it any easier for anyone. You have to have a pretty strong understanding of what is happening already to know how this is working. I am also not sure why we do not just normalize internally always. Why would we not just keep a normalized set of the scope. Is there a point of keeping the original `write_products, read_products` instead of just `write_products`. The interface is not great either. Why not just:

```ruby
orig = Scopes.new(['write_products`, 'read_products', 'read_orders'])
new = Scopes.new("write_products, read_orders")
orig == new # true
params[:scope] = new.to_s # write_products, read_orders
```

To make sure that other apps are not writing their own, we need to make our own code easy to use.

rezaansyed commented:
> Okay, so the code looks okay but I do not see how this makes it any easier for anyone. You have to have a pretty strong understanding of what is happening already to know how this is working. I am also not sure why we do not just normalize internally always. Why would we not just keep a normalized set of the scope. Is there a point of keeping the original `write_products, read_products` instead of just `write_products`. The interface is not great either. Why not just:
> 
> ```ruby
> orig = Scopes.new(['write_products`, 'read_products', 'read_orders'])
> new = Scopes.new("write_products, read_orders")
> orig == new # true
> params[:scope] = new.to_s # write_products, read_orders
> ```
> 
> To make sure that other apps are not writing their own, we need to make our own code easy to use.

@tanema you're right that we would never need to explicitly normalize the list of scopes as there is no need to keep the implicit scopes. @alexaitken and I pushed up a change to do normalize by default and also have the constructor take in either an array or string of scopes as suggested. We've renamed the value object from `Scopes` to `ApiAccess` as it encompasses the access that the scopes provide.

We also noticed that there is a need at times by apps to check whether certain scopes are available for a given access (ex: access on a token). Because of this, we added `covers?`, which will check to see whether a given list of scopes is covered by an ApiAccess.

We also wanted to ask what you thought about moving `ApiAccess` from `omniauth-shopify-oauth2` to the `shopify_api` gem? 
**Our reasoning:** 
- `ApiAccess` encapsulates what api access means.  
- This concept fits closer to using the Shopify api which doesn't require using the omniauth strategy.
- OmniAuth is using this concept to determine if the access token it retrieved as the same access that it requested.

**Cost**: this creates a new dependency in the `omniauth-shopify-oauth2`

tanema commented:
This looks amazing. The interface is great and I love `covers?`, it's a great addition.

>We also wanted to ask what you thought about moving ApiAccess from omniauth-shopify-oauth2 to the shopify_api gem?
Our reasoning:
> - ApiAccess encapsulates what api access means.
> - This concept fits closer to using the Shopify api which doesn't require using the omniauth strategy.
> - OmniAuth is using this concept to determine if the access token it retrieved as the same access that it requested.

I think these are all good reasons to move this into ShopifyAPI. The functionality is also not strictly tied to omniauth, and can be used across all of the ruby libraries (ShopifyApp's `config.scopes` for instance) It will also make more sense when we do future work on a main shopify-ruby library that will have all these functionalities in it and then omniauth and other libraries can use it. Soon we will want this sort of functionality in a central place without being tied to a framework like omniauth.

I do wonder what happens to bundler with circular dependencies but I imagine we will find out.

rezaansyed commented:
> This looks amazing. The interface is great and I love `covers?`, it's a great addition.
> 
> > We also wanted to ask what you thought about moving ApiAccess from omniauth-shopify-oauth2 to the shopify_api gem?
> > Our reasoning:
> > 
> > * ApiAccess encapsulates what api access means.
> > * This concept fits closer to using the Shopify api which doesn't require using the omniauth strategy.
> > * OmniAuth is using this concept to determine if the access token it retrieved as the same access that it requested.
> 
> I think these are all good reasons to move this into ShopifyAPI. The functionality is also not strictly tied to omniauth, and can be used across all of the ruby libraries (ShopifyApp's `config.scopes` for instance) It will also make more sense when we do future work on a main shopify-ruby library that will have all these functionalities in it and then omniauth and other libraries can use it. Soon we will want this sort of functionality in a central place without being tied to a framework like omniauth.
> 
> I do wonder what happens to bundler with circular dependencies but I imagine we will find out.

Thanks @tanema! Do you think it's better to just move this over to ShopifyAPI right now and closing this PR?

tanema commented:
>Thanks @tanema! Do you think it's better to just move this over to ShopifyAPI right now and closing this PR?

Yes
